File: 3rd Year/PPD/tema_lab_2/tema_lab_2C++/LiniiDinamic.cpp
#include "LiniiDinamic.h"
#include <fstream>

LiniiDinamic::LiniiDinamic(const int N, const int M, const int n, const int p)
    : N(N), M(M), n(n), p(p)
{
    allocate();
}

LiniiDinamic::~LiniiDinamic()
{
    deallocate();
}

void LiniiDinamic::allocate()
{
    matrix = new int*[N];
    for (int i = 0; i < N; i++)
        matrix[i] = new int[M];

    convMatrix = new int*[n];
    for (int i = 0; i < n; i++)
        convMatrix[i] = new int[n];
}

void LiniiDinamic::deallocate() const
{
    for (int i = 0; i < N; i++)
        delete[] matrix[i];
    delete[] matrix;

    for (int i = 0; i < n; i++)
        delete[] convMatrix[i];
    delete[] convMatrix;
}

void LiniiDinamic::loadData(const vector<vector<int>>& mat,
                            const vector<vector<int>>& conv) const
{
    for (int i = 0; i < N; i++)
        for (int j = 0; j < M; j++)
            matrix[i][j] = mat[i][j];
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            convMatrix[i][j] = conv[i][j];
}

void LiniiDinamic::writeToFile(const char* path) const
{
    ofstream fout(path);
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
            fout << matrix[i][j] << " ";
        fout << "\n";
    }
    fout.close();
}

void LiniiDinamic::calculateRow(const vector<int>& prev,
                                const vector<int>& current,
                                const vector<int>& next,
                                vector<int>& result) const
{
    int offset = n / 2;

    for (int col = 0; col < M; col++)
    {
        long long sum = 0;
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n; j++)
            {
                int jj = col - offset + j;
                if (jj < 0) jj = 0;
                if (jj >= M) jj = M - 1;

                int val;
                if (i < offset)
                {
                    val = prev[jj];
                }
                else if (i > offset)
                {
                    val = next[jj];
                }
                else
                {
                    val = current[jj];
                }
                sum += (long long)val * convMatrix[i][j];
            }
        }
        result[col] = (int)sum;
    }
}

void LiniiDinamic::worker(int t, int startRow, int endRow, my_barrier &barrier) {
    vector<int> prevRow(M);
    vector<int> currentRow(M);
    vector<int> nextRow(M);
    vector<int> outRow(M);

    vector<int> bufferUp(M), bufferDown(M);

    int upIdx = (startRow == 0) ? 0 : (startRow - 1);
    for (int j = 0; j < M; ++j)
        bufferUp[j] = matrix[upIdx][j];

    int downIdx = (endRow < N) ? endRow : (N - 1);
    for (int j = 0; j < M; ++j)
        bufferDown[j] = matrix[downIdx][j];

    for (int j = 0; j < M; ++j)
        prevRow[j] = bufferUp[j];
    for (int j = 0; j < M; ++j)
        currentRow[j] = matrix[startRow][j]; // prima linie din chunk

    if (startRow + 1 < endRow) {
        for (int j = 0; j < M; ++j)
            nextRow[j] = matrix[startRow + 1][j];
    } else {
        for (int j = 0; j < M; ++j)
            nextRow[j] = bufferDown[j];
    }

    barrier.wait();

    for (int row = startRow; row < endRow; ++row) {
        calculateRow(prevRow, currentRow, nextRow, outRow);
        for (int col = 0; col < M; ++col)
            matrix[row][col] = outRow[col];

        if (row + 1 < endRow) {
            // rotate window
            prevRow.swap(currentRow);
            currentRow.swap(nextRow);

            // load the new 'nextRow'
            if (row + 2 < endRow) {
                for (int col = 0; col < M; ++col)
                    nextRow[col] = matrix[row + 2][col];
            } else {
                for (int col = 0; col < M; ++col)
                    nextRow[col] = bufferDown[col];
            }
        }
    }
}


void LiniiDinamic::run()
{
    vector<thread> threads;
    my_barrier barrier(p);

    int rowsPerThread = N / p;
    int extra = N % p;
    int start = 0;

    for (int t = 0; t < p; ++t) {
        int end = start + rowsPerThread + (extra > 0 ? 1 : 0);
        if (extra > 0) extra--;

        threads.emplace_back(&LiniiDinamic::worker, this, t, start, end,
                             ref(barrier));
        start = end;
    }

    for (auto& th : threads)
        th.join();
}

File: 3rd Year/SEMESTER 1/PPD/tema_lab_2/tema_lab_2_java/tema_lab_2_java/src/LiniiDinamic.java
import java.io.FileWriter;
import java.io.IOException;
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;

public class LiniiDinamic {
    private final int N, M, n, p;
    private final int[][] matrix;
    private final int[][] convMatrix;

    public LiniiDinamic(int N, int M, int n, int p) {
        this.N = N;
        this.M = M;
        this.n = n;
        this.p = p;
        this.matrix = new int[N][M];
        this.convMatrix = new int[n][n];
    }

    public void loadData(int[][] mat, int[][] conv) {
        for (int i = 0; i < N; i++)
            System.arraycopy(mat[i], 0, matrix[i], 0, M);
        for (int i = 0; i < n; i++)
            System.arraycopy(conv[i], 0, convMatrix[i], 0, n);
    }

    public void writeToFile(String path) throws IOException {
        try (FileWriter fout = new FileWriter(path)) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++)
                    fout.write(matrix[i][j] + " ");
                fout.write("\n");
            }
        }
    }

    private void calculateRow(int[] prev, int[] current, int[] next, int[] result) {
        int offset = n / 2;

        for (int col = 0; col < M; col++) {
            long sum = 0;

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    int jj = col - offset + j;

                    if (jj < 0) jj = 0;
                    if (jj >= M) jj = M - 1;

                    int val;
                    if (i < offset)
                        val = prev[jj];
                    else if (i > offset)
                        val = next[jj];
                    else
                        val = current[jj];

                    sum += (long) val * convMatrix[i][j];
                }
            }

            result[col] = (int) sum;
        }
    }

    private void worker(int t, int startRow, int endRow, CyclicBarrier barrier) {
        int[] prevRow = new int[M];
        int[] currentRow = new int[M];
        int[] nextRow = new int[M];
        int[] outRow = new int[M];
        int[] bufferUp = new int[M];
        int[] bufferDown = new int[M];

        int upIdx = (startRow == 0) ? 0 : (startRow - 1);
        int downIdx = (endRow < N) ? endRow : (N - 1);

        System.arraycopy(matrix[upIdx], 0, bufferUp, 0, M);
        System.arraycopy(matrix[downIdx], 0, bufferDown, 0, M);
        System.arraycopy(bufferUp, 0, prevRow, 0, M);
        System.arraycopy(matrix[startRow], 0, currentRow, 0, M);

        if (startRow + 1 < endRow)
            System.arraycopy(matrix[startRow + 1], 0, nextRow, 0, M);
        else
            System.arraycopy(bufferDown, 0, nextRow, 0, M);

        try {
            barrier.await();

            for (int row = startRow; row < endRow; row++) {
                calculateRow(prevRow, currentRow, nextRow, outRow);

                System.arraycopy(outRow, 0, matrix[row], 0, M);

                if (row + 1 < endRow) {
                    int[] temp = prevRow;
                    prevRow = currentRow;
                    currentRow = nextRow;
                    nextRow = temp;

                    if (row + 2 < endRow)
                        System.arraycopy(matrix[row + 2], 0, nextRow, 0, M);
                    else
                        System.arraycopy(bufferDown, 0, nextRow, 0, M);
                }
            }
        } catch (InterruptedException | BrokenBarrierException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        Thread[] threads = new Thread[p];
        CyclicBarrier barrier = new CyclicBarrier(p);

        int rowsPerThread = N / p;
        int extra = N % p;
        int start = 0;

        for (int t = 0; t < p; ++t) {
            int end = start + rowsPerThread + (extra > 0 ? 1 : 0);
            if (extra > 0) extra--;

            final int threadId = t;
            final int s = start, e = end;
            threads[t] = new Thread(() -> worker(threadId, s, e, barrier));
            threads[t].start();

            start = end;
        }

        for (Thread th : threads) {
            try {
                th.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}
